### ISSUE TITLE
Polymer 3.0 support is broken

### ISSUE THREAD

**Issue by MajorBreakfast** [PolymerElements/marked-element#81]
- [x] `package.json` misses `marked` as a dependency
Fixed through recent PR https://github.com/Polymer/polymer-modulizer/pull/328 A new preview version of `<marked-element>` is coming up
- [x] Import of `marked` in `marked-import.js` needs two `../` (-> `../../marked/lib/marked.js'`)
- [x] When using webpack, the `marked` library does not set the global variable. This means `window.marked` is `undefined` in `marked-element.js`. This leads to an error.

**Comment by stramel** [comment]
Related to #76

You can use my branch that I used for #77. The goal is to fix the issues upstream. Most issues have been fixed upstream but not re-ran yet.

**Comment by MajorBreakfast** [comment]
@stramel Thx for the quick reply.

I've another thing: The `marked` library exports its result using CommonJS syntax when using webpack. In this case the global variable on window won't be set and consequently the code of `<marked-element>` breaks. That probably needs some kind of fix in this repo.

In any case I'd prefer if you kept the issue open until it is addressed and verified to be working. Closed issues are difficult to find and there's also a danger that it gets overlooked.

**Comment by MajorBreakfast** [comment]
👍 Thanks for reopening!

**Comment by evancaldwell** [comment]
Noticing this is still not merged, any indication as to what the status is? Also, do I need to install marked.js separately? If I remember correctly, in Polymer 1.0 that wasn't necessary.

**Comment by stramel** [comment]
We either need to find a solid solution for replacing marked library or fork marked to be es-module compliant.

**Comment by evancaldwell** [comment]
I see, is there any workaround in the meantime? I'm building some documentation for internal Polymer components and the `iron-doc-viewer` uses `marked-element`.

**Comment by MajorBreakfast** [comment]
@evancaldwell There's a workaround if you're using webpack: Install the marked package using npm and use webpack's alias feature to map `../marked` to the `marked` package. Here's how to do it:
```JS
{
  // Your webpack config
  resolve: {
    alias: {
      '../marked': 'marked'
    }
  }
}
```

Also you need to make `marked` a global variable manually somewhere before your app is loaded:
```JS
window.marked = require('marked')
```
The way marked is loaded in [marked-import.js](https://github.com/PolymerElements/marked-element/blob/3.0-preview/marked-import.js) doesn't work, because marked detects that it is loaded as a module and therefore doesn't create its global.

**Comment by evancaldwell** [comment]
@MajorBreakfast I hadn't even thought of that. It's working perfectly for `marked` but for anyone who might come looking later, because npm's `prism` is actually `prismjs` you have to:
```
{
  resolve: {
    alias: {
      '../marked': 'marked',
      '../prism': 'prismjs'
    }
  }
}
```

**Comment by MajorBreakfast** [comment]
I've edited my post above because I had forgotten to mention the global variable thing that I had to fix as well.

**Comment by evancaldwell** [comment]
@MajorBreakfast where are you putting the `window.marked = require('marked')`?

**Comment by MajorBreakfast** [comment]
At the top of my main JavaScript module (root module of my webpack bundle). Something like:
```JS
window.marked = require('marked')
require('my-app-component.js')
```

Again, I had to do this, so that the `<marked-element>` was able to find the marked library which didn't create its global variable because it detected that it was loaded into an environment with modules (webpack).

***

Side note: I'm using the good old `require()` method, because otherwise the order would be wrong (imported modules always run before the module code):
```JS
import marked from 'marked' // 1
window.marked = marked // 3

import 'my-app-component.js' // 2
```
Though, you could avoid `require()` by creating a separate module that contains just the first two lines.

**Comment by FredKSchott** [comment]
Thanks for all the feedback and triaging help everyone! We're getting ready to release a new batch of elements which should hopefully include a working marked-element, you can follow our progress here (3.0 elements are auto-generated using our polymer-modulizer tool): https://github.com/Polymer/polymer-modulizer/

**Comment by mgibas** [comment]
Issue seems to be fixed in `polymer-modulizer` (per https://github.com/Polymer/polymer-modulizer/issues/174). Should we update marked-element (rerun modulizer) ?

**Comment by MajorBreakfast** [comment]
@mgibas They do that from time to time. The last time was only 7 days ago.

The `package.json` still hasn't listed `marked` as a dependency.
https://github.com/PolymerElements/marked-element/blob/__auto_generated_3.0_preview/package.json

The other problem is fixed though! I've updated my post above to reflect this.

**Comment by e111077** [comment]
Fixed upstream see: Polymer/polymer-modulizer#328.

Will close when we run the modulizer catalog-wide again

**Comment by MajorBreakfast** [comment]
@e111077 There is still the webpack problem. `marked-import.js` should be transformed into something that works with and without webpack:

```JS
// Edit: Not working. Use the version in my next post
import marked from '../../marked/lib/marked.js'; // Produces an export in a CommonJS compatible env
if (marked) { window.marked = marked; } // Assign global variable manually
```

**Comment by mgibas** [comment]
@MajorBreakfast you are right - marked has CommonJS exports :(

`import marked from '../../marked/lib/marked.js'` works only with webpacker
`import '../../marked/lib/marked.js'` works only without webpacker

Is there any workaround to make it work in both scenarios ?

**Comment by MajorBreakfast** [comment]
@mgibas Does it break? Try this `marked-import.js` then:
```JS
import '../../marked/lib/marked.js';
if (!window.marked) { // Fix for webpack
  window.marked = require('../../marked/lib/marked.js');
}
```

**Comment by mgibas** [comment]
@MajorBreakfast ugh, so simple - that worked, thanks!

**Comment by MajorBreakfast** [comment]
I've created a PR for the webpack support.
https://github.com/PolymerElements/marked-element/pull/87

**Comment by e111077** [comment]
webpack fixed in #87 marked dep fixed in Polymer/polymer-modulizer#328
